File: dp/c/value_object.c
// Value Object Pattern in C

#include <stdio.h>
#include <string.h>
#include <stdbool.h>

typedef struct { long amount; char currency[4]; } Money;

Money money_new(long cents, const char* currency) {
    Money m;
    m.amount = cents;
    snprintf(m.currency, 4, "%s", currency);
    return m;
}

Money money_dollars(double d) { return money_new((long)(d * 100), "USD"); }

Money money_add(Money a, Money b) {
    if (strcmp(a.currency, b.currency) != 0) {
        printf("ERROR: cannot add %s + %s\n", a.currency, b.currency);
        return a;
    }
    return money_new(a.amount + b.amount, a.currency);
}

Money money_multiply(Money m, long factor) {
    return money_new(m.amount * factor, m.currency);
}

bool money_equal(Money a, Money b) {
    return a.amount == b.amount && strcmp(a.currency, b.currency) == 0;
}

void money_print(Money m) {
    printf("%ld.%02ld %s", m.amount / 100, m.amount % 100, m.currency);
}

typedef struct { char street[64]; char city[32]; char country[4]; } Address;

bool address_equal(const Address* a, const Address* b) {
    return strcmp(a->street, b->street) == 0 && strcmp(a->city, b->city) == 0
        && strcmp(a->country, b->country) == 0;
}

int main(void) {
    Money p1 = money_dollars(29.99);
    Money p2 = money_dollars(10.01);
    Money total = money_add(p1, p2);

    money_print(p1); printf(" + "); money_print(p2);
    printf(" = "); money_print(total); printf("\n");

    Money a = money_dollars(10.0);
    Money b = money_dollars(10.0);
    printf("$10 == $10: %s\n", money_equal(a, b) ? "true" : "false");

    Money doubled = money_multiply(a, 2);
    printf("Doubled: "); money_print(doubled); printf("\n");

    Address addr1 = {"123 Main St", "Springfield", "US"};
    Address addr2 = {"456 Oak Ave", "Springfield", "US"};
    printf("Same address: %s\n", address_equal(&addr1, &addr2) ? "true" : "false");
    return 0;
}

File: dp/rust/src/value_object.rs
// Value Object Pattern in Rust
// Immutable objects defined by their attributes, not identity

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
struct Money {
    amount: i64,       // cents to avoid floating-point
    currency: String,
}

impl Money {
    fn new(amount: i64, currency: &str) -> Self {
        Money {
            amount,
            currency: currency.to_uppercase(),
        }
    }

    fn dollars(amount: f64) -> Self {
        Money::new((amount * 100.0) as i64, "USD")
    }

    fn add(&self, other: &Money) -> Result<Money, String> {
        if self.currency != other.currency {
            return Err(format!("Cannot add {} to {}", self.currency, other.currency));
        }
        Ok(Money::new(self.amount + other.amount, &self.currency))
    }

    fn multiply(&self, factor: i64) -> Money {
        Money::new(self.amount * factor, &self.currency)
    }

    fn display_amount(&self) -> String {
        format!("{}.{:02} {}", self.amount / 100, (self.amount % 100).abs(), self.currency)
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
struct Address {
    street: String,
    city: String,
    country: String,
    postal_code: String,
}

impl Address {
    fn new(street: &str, city: &str, country: &str, postal_code: &str) -> Self {
        Address {
            street: street.to_string(),
            city: city.to_string(),
            country: country.to_string(),
            postal_code: postal_code.to_string(),
        }
    }

    fn with_street(&self, street: &str) -> Self {
        Address {
            street: street.to_string(),
            ..self.clone()
        }
    }
}

fn main() {
    // Money value objects
    let price1 = Money::dollars(29.99);
    let price2 = Money::dollars(10.01);
    let total = price1.add(&price2).unwrap();
    println!("{} + {} = {}", price1.display_amount(), price2.display_amount(), total.display_amount());

    // Equality by value
    let a = Money::dollars(10.0);
    let b = Money::dollars(10.0);
    println!("$10 == $10: {}", a == b); // true (same value)

    // Immutability: multiply returns new object
    let doubled = a.multiply(2);
    println!("Original: {}, Doubled: {}", a.display_amount(), doubled.display_amount());

    // Address value object
    let addr = Address::new("123 Main St", "Springfield", "US", "62701");
    let moved = addr.with_street("456 Oak Ave");
    println!("\nOriginal: {:?}", addr);
    println!("New addr: {:?}", moved);
    println!("Same? {}", addr == moved); // false (different street)
}
